File: app/src/main/java/com/mailerdaemon/app/Notices/NoticesActivity.java
package com.mailerdaemon.app.Notices;

import android.os.Bundle;
import android.preference.PreferenceManager;
import android.view.MenuItem;
import android.view.View;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import com.facebook.shimmer.ShimmerFrameLayout;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.google.common.collect.Lists;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.Query;
import com.google.firebase.firestore.Source;
import com.google.gson.Gson;
import com.mailerdaemon.app.R;

import java.util.List;
import java.util.Objects;

import Utils.AccessDatabse;
import Utils.DialogOptions;
import Utils.StringRes;

public class NoticesActivity extends AppCompatActivity implements AccessDatabse, DialogOptions {


    private ShimmerFrameLayout shimmerViewContainer;
    private List<DocumentSnapshot> noticeModels;
    private RecyclerView recyclerView;
    private NoticeRecyclerViewAdapter adapter;
    private FloatingActionButton fab_add;
    private Boolean access;
    private SwipeRefreshLayout refresh;

    @Override
    public void onCreate( Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_notices);
        Objects.requireNonNull(getSupportActionBar()).setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setTitle("Notices");
        access= PreferenceManager.getDefaultSharedPreferences(getApplicationContext()).getBoolean("Access",false);
        fab_add=findViewById(R.id.fab);
        refresh=findViewById(R.id.refresh);
        if (access){
            fab_add.show();
        }
        shimmerViewContainer = findViewById(R.id.shimmer_view_container);

        recyclerView = findViewById(R.id.rv_notices);
        adapter = new NoticeRecyclerViewAdapter( this,this);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        if (noticeModels == null){
            getDatabase();
        }
        else{
            adapter.setData(Lists.reverse(noticeModels));
            recyclerView.setAdapter(adapter);
            adapter.notifyDataSetChanged();
        }
        fab_add.setOnClickListener(v->{
            AddNoticeFragment fragment=new AddNoticeFragment();
            fragment.show(getSupportFragmentManager(),null);
        });
        refresh.setOnRefreshListener(() -> {
            getDatabase();
            refresh.setRefreshing(false);
        });
    }


  @Override
  public void getDatabase() {
        recyclerView.setVisibility(View.INVISIBLE);
      shimmerViewContainer.setVisibility(View.VISIBLE);
      shimmerViewContainer.startShimmer();
      FirebaseFirestore.getInstance().collection(StringRes.FB_Collec_Notice).orderBy("date", Query.Direction.DESCENDING).limit(15).get(Source.SERVER).addOnCompleteListener(task -> {
      recyclerView.setVisibility(View.VISIBLE);
      if(task.isSuccessful()) {
        noticeModels = Objects.requireNonNull(task.getResult()).getDocuments();
        adapter.setData(noticeModels);
        recyclerView.setAdapter(adapter);
        adapter.notifyDataSetChanged();
        shimmerViewContainer.stopShimmer();
        shimmerViewContainer.setVisibility(View.GONE);
      }else{
//          Toast.makeText(getContext(),StringRes.No_Internet,Toast.LENGTH_LONG).show();
          FirebaseFirestore.getInstance().collection(StringRes.FB_Collec_Notice).orderBy("date", Query.Direction.DESCENDING).limit(15).get(Source.CACHE).addOnCompleteListener(task2 -> {
              if(task.isSuccessful()) {
                  noticeModels = Objects.requireNonNull(task2.getResult()).getDocuments();
                  adapter.setData(noticeModels);
                  adapter.notifyDataSetChanged();
                  shimmerViewContainer.stopShimmer();
                  shimmerViewContainer.setVisibility(View.GONE);
              }
          });
      }
    });
      recyclerView.scrollToPosition(0);
  }

    @Override
    public void showOptions(NoticeModel model,String path) {
        Bundle bundle=new Bundle();
        bundle.putString("path", path);
        bundle.putString("model", new Gson().toJson(model));
        OptionsFragment optionsFragment=new OptionsFragment();
        optionsFragment.setArguments(bundle);
        optionsFragment.show(getSupportFragmentManager(), null);

    }

    @Override
    public void showDialog(String path) {

    }
    @Override
    public boolean onOptionsItemSelected(MenuItem item) {

        if (item.getItemId() == android.R.id.home)
            onBackPressed();
        return true;
    }
}

File: app/src/main/java/com/mailerdaemon/app/notices/NoticesActivity.kt
package com.mailerdaemon.app.notices

import android.os.Bundle
import android.view.MenuItem
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import com.mailerdaemon.app.*
import kotlinx.android.synthetic.main.activity_notices.refresh
import kotlinx.android.synthetic.main.shimmer_layout_posts.*
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class NoticesActivity : AppCompatActivity() {

    var data = emptyList<PostModel>()
    private val fragmentNotices = NoticesFragment()

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_placement)
        supportActionBar?.let {
            it.setDisplayHomeAsUpEnabled(true)
            it.title = resources.getString(R.string.title_activity_notices)
        }
        supportFragmentManager.beginTransaction().add(R.id.container, fragmentNotices).commit()
    }

    fun getNotices(showNotices: ShowNotices) {
        shimmer_view_container.let {
            it.startShimmer()
            it.visibility = View.VISIBLE
        }

        (application as ApplicationClass).repository.getPosts()
            ?.enqueue(object : Callback<List<PostModel>?> {
                override fun onFailure(call: Call<List<PostModel>?>, t: Throwable) {
                    baseContext.toast("Error")
                }

                override fun onResponse(
                    call: Call<List<PostModel>?>,
                    response: Response<List<PostModel>?>
                ) {
                    val result = response.body()
                    if (response.isSuccessful && result != null) {
                        if (shimmer_view_container.isShimmerVisible) {
                            shimmer_view_container.let {
                                it.stopShimmer()
                                it.visibility = View.GONE
                            }
                        }
                        refresh.visibility = View.VISIBLE
                        var list = mutableListOf<PostModel>()
                        for (element in result) {
                            if (!isSomethingNull(element)) {
                                list.add(element)
                            }
                        }
                        data = list
                        showNotices.showNotices(data)
                    } else
                        baseContext.toast("Error")
                }
            })
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) onBackPressed()
        return true
    }

    companion object {
        interface ShowNotices {
            fun showNotices(list: List<PostModel>)
        }

        const val noticeData = "notices"
    }

    fun isSomethingNull(element: PostModel): Boolean {
        return element._id == null || element.created_time == null
                || element.id == null || element.message == null || element.permalink_url == null
    }
}
